File: cmd/open/action.go
package open

import (
	"errors"

	"github.com/fioncat/roxide/cmd"
	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/repoutils"
	"github.com/fioncat/roxide/pkg/term"
	"github.com/spf13/cobra"
)

func newAction() *cobra.Command {
	var opts actionOptions

	c := &cobra.Command{
		Use:   "action",
		Short: "Open action",

		Args: cobra.NoArgs,

		ValidArgsFunction: cmd.NoneCompletion,
	}

	return cmd.Build(c, &opts)
}

type actionOptions struct{}

func (o *actionOptions) Complete(c *cobra.Command, args []string) error {
	return nil
}

func (o *actionOptions) Run(ctx *context.Context) error {
	term.Mute = true
	repo, err := repoutils.MustGetCurrentRepo(ctx)
	if err != nil {
		return err
	}
	err = ctx.SetRepo(repo)
	if err != nil {
		return err
	}

	req, err := repoutils.GetActionRequest(ctx)
	if err != nil {
		return err
	}

	api, err := ctx.RemoteAPI(repo.Remote)
	if err != nil {
		return err
	}

	action, err := api.GetAction(req)
	if err != nil {
		return err
	}

	if action == nil {
		return errors.New("no action found")
	}

	if action.URL == "" {
		return errors.New("this action has no url to open")
	}

	return term.OpenURL(action.URL)
}

File: src/cmd/open/action.rs
use anyhow::{Context, Result};
use async_trait::async_trait;
use clap::Args;

use crate::cmd::Command;
use crate::cmd::complete::CompleteCommand;
use crate::config::context::ConfigContext;
use crate::debug;
use crate::repo::current::get_current_repo;
use crate::repo::wait_action::WaitActionArgs;
use crate::term::list::TableArgs;

/// Open current action in the browser.
#[derive(Debug, Args)]
pub struct OpenActionCommand {
    #[clap(flatten)]
    pub wait_action: WaitActionArgs,

    #[clap(flatten)]
    pub table: TableArgs,
}

#[async_trait]
impl Command for OpenActionCommand {
    fn name() -> &'static str {
        "action"
    }

    async fn run(self, mut ctx: ConfigContext) -> Result<()> {
        debug!("[cmd] Run open action command: {:?}", self);
        ctx.mute();

        let repo = get_current_repo(&ctx)?;
        let api = ctx.get_api(&repo.remote, false)?;

        let action = self.wait_action.wait(&ctx, &repo, api.as_ref()).await?;

        open::that(&action.web_url)
            .with_context(|| format!("failed to open action: {}", action.web_url))
    }

    fn complete() -> CompleteCommand {
        Self::default_complete()
            .arg(WaitActionArgs::complete())
            .args(TableArgs::complete())
    }
}
